provider/etcd/etcd.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2016 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcd
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/alertmanager/provider"
+	"github.com/prometheus/alertmanager/store"
+	"github.com/prometheus/alertmanager/types"
+)
+
+const alertChannelLength = 200
+
+// Alerts gives access to a set of alerts. All methods are goroutine-safe.
+type Alerts struct {
+	cancel context.CancelFunc
+
+	mtx       sync.Mutex
+	alerts    *store.Alerts
+	listeners map[int]listeningAlerts
+	next      int
+
+	logger log.Logger
+
+	EtcdClient *EtcdClient
+}
+
+type listeningAlerts struct {
+	alerts chan *types.Alert
+	done   chan struct{}
+}
+
+// NewAlerts returns a new alert provider.
+func NewAlerts(ctx context.Context, m types.Marker, intervalGC time.Duration, l log.Logger,
+	etcdEndpoints []string, etcdPrefix string) (*Alerts, error) {
+
+	ctx, cancel := context.WithCancel(ctx)
+	a := &Alerts{
+		alerts:    store.NewAlerts(intervalGC),
+		cancel:    cancel,
+		listeners: map[int]listeningAlerts{},
+		next:      0,
+		logger:    log.With(l, "component", "provider"),
+	}
+	a.alerts.SetGCCallback(func(alerts []*types.Alert) {
+		for _, alert := range alerts {
+			// As we don't persist alerts, we no longer consider them after
+			// they are resolved. Alerts waiting for resolved notifications are
+			// held in memory in aggregation groups redundantly.
+			fp := alert.Fingerprint()
+			m.Delete(fp)
+			a.EtcdClient.Del(fp)
+		}
+
+		a.mtx.Lock()
+		for i, l := range a.listeners {
+			select {
+			case <-l.done:
+				delete(a.listeners, i)
+				close(l.alerts)
+			default:
+				// listener is not closed yet, hence proceed.
+			}
+		}
+		a.mtx.Unlock()
+	})
+	a.alerts.Run(ctx)
+
+	// initialize etcd client
+	etcdClient, err := NewEtcdClient(ctx, a, etcdEndpoints, etcdPrefix)
+	if err != nil {
+		return nil, err
+	}
+	a.EtcdClient = etcdClient
+	return a, nil
+}
+
+// Close the alert provider.
+func (a *Alerts) Close() {
+	if a.cancel != nil {
+		a.cancel()
+	}
+}
+
+func max(a, b int) int {
```

Comment:
SHOULD: Add a function comment here. (pedantic, I know)

Reply:
that's not my code.  Trying to leave it as similar to mem.go as possible.